### PR TITLE
Clean up timing of layout tree flattening/ copying of unflattened tree for Weaver

### PR DESCRIPTION
**Pull request by mikezucc**
Before display, a node's layout tree is flattened in order to calculate display information such as position. In almost all cases, this happens early in the layout transition lifecycle (during `calculateLayoutThatFits:` ). However, there are special cases where this layout tree is not flattened until the very last opportunity so that debugging tools (Weaver) can inspect the layout tree.

With the move to use const references instead of shared_ptr's for the pending layout tree used in the layout transition, the flattened layout tree was accessible to changes in the pending stage of the transition, causing layout specs to sneak into the final step of the layout transition: moving subnodes.

Huy simplified this diff to just a few lines.

### CODE REVIEW

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Any source code changes should have an entry in CHANGELOG.md or have #trivial in their title.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: Source/ASDisplayNode+Layout.mm [hunk 1]
```diff
@@ -993,12 +993,6 @@ - (void)_pendingLayoutTransitionDidComplete
   _pendingLayoutTransition = nil;
 }
 
-- (void)_setCalculatedDisplayNodeLayout:(const ASDisplayNodeLayout &)displayNodeLayout
```
Comment (nguyenhuy) [inline Source/ASDisplayNode+Layout.mm]: For the record, this method is not valled anywhere.

File: Source/ASDisplayNode.mm [hunk 2]
```diff
@@ -1220,11 +1220,10 @@ - (ASLayout *)calculateLayoutThatFits:(ASSizeRange)constrainedSize
   }
   ASDisplayNodeLogEvent(self, @"computedLayout: %@", layout);
 
-  // Return the (original) unflattened layout if it needs to be stored. The layout will be flattened later on (@see _locked_setCalculatedDisplayNodeLayout:).
-  // Otherwise, flatten it right away.
-  if (! [ASDisplayNode shouldStoreUnflattenedLayouts]) {
-    layout = [layout filteredNodeLayoutTree];
+  if ([ASDisplayNode shouldStoreUnflattenedLayouts]) {
+      _unflattenedLayout = layout;
```
Comment (nguyenhuy) [inline Source/ASDisplayNode.mm]: As discussed offline, there is a small risk that the layout here will not be used/applied later on, and Weaver (or any other users of this debugging feature) will end up with a wrong unflattened layout. But given it's a debugging feature and should not be used in producation, the risk is small and doesn't warrant the complexity in previous implementation.

Comment (maicki) [reply Source/ASDisplayNode.mm]: Can you add some comment about this fact?

**Comment by mikezucc** [comment]
I will rebase with the correct author, no worry

**Pull request merged**